java-checks/src/main/java/org/sonar/java/checks/RedundantConstructorsAndMethodsShouldBeAvoidedCheck.java

```diff
@@ -0,0 +1,147 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionStatementTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S6207")
+public class RedundantConstructorsAndMethodsShouldBeAvoidedCheck extends IssuableSubscriptionVisitor {
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.RECORD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree targetRecord = (ClassTree) tree;
+
+    List<VariableTree> components = targetRecord.recordComponents();
+    Map<String, VariableTree> componentsByName = components.stream()
+      .collect(Collectors.toMap(component -> component.symbol().name(), component -> component));
+
+    for (Tree member : targetRecord.members()) {
+      if (member.is(Tree.Kind.CONSTRUCTOR)) {
+        MethodTree constructor = (MethodTree) member;
+        // Report if the constructor is empty
+        if (constructor.block().body().isEmpty() || onlyDoesSimpleAssignments(constructor, components)) {
+          reportIssue(constructor.simpleName(), "Remove this redundant constructor which is the same as a default one.");
+        }
+      } else if (member.is(Tree.Kind.METHOD)) {
+        MethodTree method = (MethodTree) member;
+        String methodName = method.symbol().name();
+        if (!componentsByName.containsKey(methodName)) {
+          continue;
+        }
+        if (onlyReturnsRawValue(method, components)) {
+          reportIssue(method.simpleName(), "Remove this redundant method which is the same as a default one.");
+        }
+      }
+    }
+  }
+
+  public static boolean onlyReturnsRawValue(MethodTree method, Collection<VariableTree> components) {
+    List<ReturnStatementTree> returnStatements = method.block().body().stream()
```

Comment:
I feel this should be done simpler. Also, I don't see any test with more than one statement, so I don't know if the loop is working fine. To me, it is suspcisious.

Try using something similar as in: 

https://github.com/SonarSource/sonar-java/blob/master/java-checks/src/main/java/org/sonar/java/checks/GettersSettersOnRightFieldCheck.java#L135

Also, I try to achieve the same thing on my PR, here: https://github.com/SonarSource/sonar-java/pull/3685/files#diff-98bea908beb6ebd2a216e7e338702537b1ff7479f297b953001cd1ede773383fR79